### PR TITLE
Add an actually synchronous method to IThreadHandling

### PR DESCRIPTION
**Pull request by bigfluffycookie**
fixes https://github.com/SonarSource/sonarlint-visualstudio/issues/4415

### CODE REVIEW

File: src/TestInfrastructure/NoOpThreadHandler.cs [hunk 1]
```diff
@@ -46,10 +46,14 @@ public Task RunOnUIThread(Action op)
 
         public void RunOnUIThreadSync(Action op) => op();
 
+        public void RunOnUIThreadSync2(Action op) => op();
+
         public Task<T> RunOnBackgroundThread<T>(Func<Task<T>> asyncMethod) => asyncMethod();
 
         public IAwaitableWrapper SwitchToBackgroundThread() => new NoOpAwaitable();
 
+        public Task SwitchToMainThreadAsync() => Task.CompletedTask;
```
Comment (bigfluffycookie) [inline src/TestInfrastructure/NoOpThreadHandler.cs:55]: This was on a separate commit than the original SwitchToMainThreadAsync add-in was so it was simpler for me to just add in this line. (I made sure it is in the same place.)

**Comment by bigfluffycookie** [comment]
One code smell is about not blocking the thread, I assume I can ignore that as we want to block? Then two of them are about the obsolete, should I have written a comment about why it is obsolete? (I could link to the but ticket or something)

File: src/Core/IThreadHandling.cs [hunk 2]
```diff
@@ -61,8 +61,18 @@ public interface IThreadHandling
         /// If the caller is not on the main thread then the method will switch to the main thread,
         /// then resume on the caller's thread when then the operation completes.
         /// </summary>
+        [Obsolete]
```
Comment (duncanp-sonar) [inline src/Core/IThreadHandling.cs:64]: Could you add a comment explaining why this is obsolete and what the alternative is, please? e.g.

```
[Obsolete("This method does not work correctly. Use RunOnUIThreadSync2 instead.")]
// See https://github.com/SonarSource/sonarlint-visualstudio/issues/4179 for more info.
```

I think the message text in the attribute might show up in IntelliSense, but I'm not sure.

File: src/Infrastructure.VS/ThreadHandling.cs [hunk 3]
```diff
@@ -52,6 +53,22 @@ public class ThreadHandling : IThreadHandling
...
+                op();
+                return;
+            }
+            ThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                op();
+            });
+
+            Debug.Assert(!ThreadHelper.CheckAccess(), "Not expecting to returning on the UI thread");
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/ThreadHandling.cs:69]: Minor: missing word  - "...be returning..."

(or just "...to be on...")

**Review by duncanp-sonar — APPROVED** [review]
Looks good. Just a suggestion about commenting the attribute

**Pull request merged**